llvm/lib/DebugInfo/DWARF/DWARFTypePrinter.cpp

```diff
@@ -173,7 +169,8 @@ DWARFTypePrinter::appendUnqualifiedNameBefore(DWARFDie D,
   case DW_TAG_base_type:
   */
   default: {
-    const char *NamePtr = dwarf::toString(D.find(DW_AT_name), nullptr);
+    const char *NamePtr =
+        dwarf::toString(D.findRecursively(DW_AT_name), nullptr);
```

Comment:
possibly if D is resolved through type unit references before this line, then this findRecursively won't be needed and the resolve through type units won't be needed in appendTemplateParameters either?

Reply:
That would work, but I think that basically amounts to reverting all the changes in this file (as that's what the original code was doing). Is that what you're suggesting?

(The reason I changed this file was to demonstrate how one can write code to access the type mostly without caring about type units. It's not really required to make the main change (have recursive resolution look through type unit references), but it makes me wonder whether it is a good idea.)

Reply:
Ah, I guess maybe it'd make sense to split this patch up. (haven't thought about whether these two changes are ordered or order independent)

The `getAttributeValueAsReferencedDie` accounting for type units is one thing, sounds good.

The "recurse through DW_AT_signature" to resolve some lookups is a separate thing - I think it still does make sense... maybe? It is a bit less efficient to have to lookup type units multiple times, compared to the abstract_origin/specification indirection which is at least just jumping to an offset (though I guess it's still a binary search through a list of DIEs, ultimately, to find the DIE with a given offset?).

I guess alternatively, we could make the children iteration do the same indirection too? Though doing that implicitly might be too much (like even the dumper would get confused, dumping the children of the skeleton type DIE, etc) so then it'd have to be a separate function... 

I guess maybe my thoughts are that findRecursively, finding recursively explicitly is OK - but that maybe it doesn't suit the use case here in the type printer (because it's comprehensively visiting lots of stuff - whereas just "give me the name of this thing" or "give me the file/line this thing was defined on" is a relatively more common/mild query that's nice to be able to do more conveniently)

Hmm - when/where/how does /this/ function (`appendUnqualifiedNameBefore`) get called on an unresolved type? Perhaps this should be pushed further up... like every call to "get a DIE from this DW_AT_type" is already looking through to the type unit, what's left/how else does this code end up dealing with a skeleton type unit? (I guess maybe in top level callers - perhaps that needs an explicit entry point that resolves before going into the generic walking code)

Reply:
> Ah, I guess maybe it'd make sense to split this patch up. (haven't thought about whether these two changes are ordered or order independent)
> 
> The `getAttributeValueAsReferencedDie` accounting for type units is one thing, sounds good.

Sounds good. I'll create a patch for that. In fact, I think lldb already has this part.

> 
> The "recurse through DW_AT_signature" to resolve some lookups is a separate thing - I think it still does make sense... maybe? It is a bit less efficient to have to lookup type units multiple times, compared to the abstract_origin/specification indirection which is at least just jumping to an offset (though I guess it's still a binary search through a list of DIEs, ultimately, to find the DIE with a given offset?).

It is, and I guess that the direct lookup would always be faster, though maybe not it a way that we care.

> I guess alternatively, we could make the children iteration do the same indirection too? Though doing that implicitly might be too much (like even the dumper would get confused, dumping the children of the skeleton type DIE, etc) so then it'd have to be a separate function...

Yeah, that sound like too much. We don't do that for DW_AT_specification either.

> I guess maybe my thoughts are that findRecursively, finding recursively explicitly is OK - but that maybe it doesn't suit the use case here in the type printer (because it's comprehensively visiting lots of stuff - whereas just "give me the name of this thing" or "give me the file/line this thing was defined on" is a relatively more common/mild query that's nice to be able to do more conveniently)

I can definitely understand that position.

> 
> Hmm - when/where/how does /this/ function (`appendUnqualifiedNameBefore`) get called on an unresolved type? Perhaps this should be pushed further up... like every call to "get a DIE from this DW_AT_type" is already looking through to the type unit, what's left/how else does this code end up dealing with a skeleton type unit? (I guess maybe in top level callers - perhaps that needs an explicit entry point that resolves before going into the generic walking code)

I think this can happen anywhere where we have a double reference (so when a DW_AT_type points to a declaration, which points (via DW_AT_signature) to a type unit) (*). `getAttributeValueAsReferencedDie` will only resolve the first reference, and not the second one. Unless we want to teach it to resolve references recursively (which sounds like a bad idea), we're left with explicitly calling some type-unit resolving function at every call site, which sounds a lot like what `resolveReferencedType` already is.

So maybe the verdict is:
- getAttributeValueAsReferencedDie should resolve type unit references
- findRecursively should do that as well
- we should leave type printer alone?

(*) This also means that this problem doesn't occur with gcc-style type unit references, which I'm really starting to like. :)

Reply:
Aside: GCC doesn't use this style for all type unit references, it actually uses 3 different styles depending on context
https://godbolt.org/z/Pozc4TYM1
t1, type unit referenced once `DW_AT_type [DW_FORM_ref_sig8]	(0xf8473fd3bfd1e564)`
t2, type unit referenced twice (save space by sharing the sig8): `DW_AT_type [DW_FORM_ref4]	(cu + 0x00cc => {0x000000cc} "t2")`
```
0x000000cc:   DW_TAG_structure_type [14]   (0x0000000c)
                DW_AT_signature [DW_FORM_ref_sig8]	(0x3fea40f8df132361)
```
t3: referenced a declaration of a member function:
```
0x000000b3:   DW_TAG_subprogram [13]   (0x0000000c)
                DW_AT_specification [DW_FORM_ref4]	(cu + 0x0049 => {0x00000049} "_ZN2ns2t33memEv")
                DW_AT_decl_line [DW_FORM_data1]	(9)
                DW_AT_decl_column [DW_FORM_data1]	(6)
                DW_AT_low_pc [DW_FORM_addr]	(0x0000000000000000 ".text")
                DW_AT_high_pc [DW_FORM_data8]	(0x0000000000000007)
                DW_AT_frame_base [DW_FORM_exprloc]	(DW_OP_call_frame_cfa)
                DW_AT_call_all_calls [DW_FORM_flag_present]	(true)
```
```
0x0000002e:   DW_TAG_namespace [10] * (0x0000000c)
                DW_AT_name [DW_FORM_string]	("ns")
                DW_AT_decl_file [DW_FORM_data1]	("/app/example.cpp")
                DW_AT_decl_line [DW_FORM_data1]	(1)
                DW_AT_decl_column [DW_FORM_data1]	(11)
                DW_AT_sibling [DW_FORM_ref4]	(cu + 0x0086 => {0x00000086})

0x00000039:     DW_TAG_structure_type [11] * (0x0000002e)
                  DW_AT_name [DW_FORM_string]	("t3")
                  DW_AT_declaration [DW_FORM_flag_present]	(true)
                  DW_AT_signature [DW_FORM_ref_sig8]	(0x55d7e46295b56ec8)
                  DW_AT_sibling [DW_FORM_ref4]	(cu + 0x0056 => {0x00000056})

0x00000049:       DW_TAG_subprogram [5]   (0x00000039)
                    DW_AT_external [DW_FORM_flag_present]	(true)
                    DW_AT_name [DW_FORM_string]	("mem")
                    DW_AT_decl_file [DW_FORM_data1]	("/app/example.cpp")
                    DW_AT_decl_line [DW_FORM_data1]	(4)
                    DW_AT_decl_column [DW_FORM_data1]	(25)
                    DW_AT_linkage_name [DW_FORM_strp]	( .debug_str[0x00000091] = "_ZN2ns2t33memEv")
                    DW_AT_declaration [DW_FORM_flag_present]	(true)
```
In any case... 

It seems to me that perhaps we could/should introduce a `getAttributeValueAsReferencedDieRecursively`? That recurses the referenced type, but doesn't recurse the attribute (since this is called on an already found attribute - with find or findRecursively)?

> getAttributeValueAsReferencedDie will only resolve the first reference, and not the second one. Unless we want to teach it to resolve references recursively (which sounds like a bad idea)

Hmm, you mentioned that this sounds like a bad idea, but then mention that we should do it a few lines later? Not sure which you're thinking of, or why it might be a bad idea? (I mean, it might be, not suggesting it isn't, but trying to understand better)

Or perhaps you were presenting these as alternatives...

Reply:
> Aside: GCC doesn't use this style for all type unit references, it actually uses 3 different styles depending on context

Okay, not so nice then, I guess :)


> It seems to me that perhaps we could/should introduce a `getAttributeValueAsReferencedDieRecursively`? That recurses the referenced type, but doesn't recurse the attribute (since this is called on an already found attribute - with find or findRecursively)?

This seems a bit too magical to me. `getAttributeValueAsReferencedDieRecursively()` is not that much shorter than `getAttributeValueAsReferencedDie().resolveTypeUnitReference()`, but the latter is clearer in what's happening (and I think this will need to be evaluated on a case-by-case basis, as depending on what exactly you want to do (e.g. which children to look at) you might or might not want to do the "recursive" part).

I think it would make sense if it were named somewhat differently to so that it's clear this is used for types and that it will point you to some kind of a canonical type definition... oops, I think i've just reinvented the existing
```
static DWARFDie resolveReferencedType(DWARFDie D, DWARFFormValue F) {
  return D.getAttributeValueAsReferencedDie(F).resolveTypeUnitReference();
}
```

> 
> > getAttributeValueAsReferencedDie will only resolve the first reference, and not the second one. Unless we want to teach it to resolve references recursively (which sounds like a bad idea)
> 
> Hmm, you mentioned that this sounds like a bad idea, but then mention that we should do it a few lines later? Not sure which you're thinking of, or why it might be a bad idea? (I mean, it might be, not suggesting it isn't, but trying to understand better)
> 
> Or perhaps you were presenting these as alternatives...

I'm sorry, I was being imprecise here. I think that the *recursive* reference resolution is a bad idea. The rest of the text omits the word "recursive", which was meant to mean that it should handle a single DW_FORM_ref_sig8 reference (and nothing more)

> -    getAttributeValueAsReferencedDie should resolve type unit references

It should resolve a single DW_FORM_ref_sig8 like it does in this patch

> -    findRecursively should do that as well

It should recurse on DW_AT_signature (like it does in this patch), by calling getAttributeValueAsReferencedDie

> -    we should leave type printer alone?

just that

Reply:
> > It seems to me that perhaps we could/should introduce a `getAttributeValueAsReferencedDieRecursively`? That recurses the referenced type, but doesn't recurse the attribute (since this is called on an already found attribute - with find or findRecursively)?
> 
> This seems a bit too magical to me. `getAttributeValueAsReferencedDieRecursively()` is not that much shorter than `getAttributeValueAsReferencedDie().resolveTypeUnitReference()`, but the latter is clearer in what's happening (and I think this will need to be evaluated on a case-by-case basis, as depending on what exactly you want to do (e.g. which children to look at) you might or might not want to do the "recursive" part).
> 
> I think it would make sense if it were named somewhat differently to so that it's clear this is used for types and that it will point you to some kind of a canonical type definition... oops, I think i've just reinvented the existing
> 
> ```
> static DWARFDie resolveReferencedType(DWARFDie D, DWARFFormValue F) {
>   return D.getAttributeValueAsReferencedDie(F).resolveTypeUnitReference();
> }
> ```

Fair enough - though if there's more places that want this functionality, I'd be fine with elevating it to some more general purpose helper on DWARFDie rather than hidden as a static helper in DWARFTypePrinter.

> > > getAttributeValueAsReferencedDie will only resolve the first reference, and not the second one. Unless we want to teach it to resolve references recursively (which sounds like a bad idea)
> > 
> > 
> > Hmm, you mentioned that this sounds like a bad idea, but then mention that we should do it a few lines later? Not sure which you're thinking of, or why it might be a bad idea? (I mean, it might be, not suggesting it isn't, but trying to understand better)
> > Or perhaps you were presenting these as alternatives...
> 
> I'm sorry, I was being imprecise here. I think that the _recursive_ reference resolution is a bad idea. The rest of the text omits the word "recursive", which was meant to mean that it should handle a single DW_FORM_ref_sig8 reference (and nothing more)

Ah, OK.

> > * getAttributeValueAsReferencedDie should resolve type unit references
> 
> It should resolve a single DW_FORM_ref_sig8 like it does in this patch
> 
> > * findRecursively should do that as well
> 
> It should recurse on DW_AT_signature (like it does in this patch), by calling getAttributeValueAsReferencedDie
> 
> > * we should leave type printer alone?
> 
> just that

Yeah, that sounds OK to me. Thanks for your patience/giving this a go! (& if you are touching the lldb side of this sort of change too - it'd be handy to have similar type resolution helpers (ideally /really/ similar, because Real Soon Now, we'd like the DWARFTypePrinter to be generalized across both DWARFDie and DWARFDIE), so thanks for keeping this in mind (: )

Reply:
This is ready for review now. I'm sorry it took me so long to get back to this.